File: python/techai_webutils/src/techai_webutils/core/interfaces/tracer.py
"""Tracer interface for distributed tracing.

Defines the abstract contract for tracer implementations.
Foundation implementations satisfy this interface; consumers depend only on it.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from contextlib import contextmanager
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Generator


class TracerProvider(ABC):
    """Abstract tracer for distributed tracing.

    Implementations: OpenTelemetry (default).
    All service code depends on this interface, never on a concrete tracing library.
    """

    @abstractmethod
    @contextmanager
    def span(self, name: str, **attributes: str | float | bool) -> Generator[TracerSpan, None, None]:
        """Start a new span as a context manager.

        Usage::

            with tracer.span("operation", key="value") as span:
                span.set_attribute("result", 42)
        """
        ...

    @abstractmethod
    def shutdown(self) -> None:
        """Flush and shut down the tracer."""
        ...


class TracerSpan(ABC):
    """Abstract span within a trace."""

    @abstractmethod
    def set_attribute(self, key: str, value: str | float | bool) -> None:  # noqa: FBT001
        """Add a key-value attribute to the span."""
        ...

    @abstractmethod
    def record_error(self, error: BaseException) -> None:
        """Record an error on the span."""
        ...

File: go/core/interfaces/tracer.go
package interfaces

import "context"

// Tracer provides distributed tracing capabilities.
//
// Implementations: OpenTelemetry (default), or any tracing backend
// that supports span creation and context propagation.
//
// All foundation and service code depends on this interface, never on a
// concrete tracing library. Swap implementations via the tracer.New factory.
type Tracer interface {
	// Start creates a new span and returns the updated context and span.
	Start(ctx context.Context, name string, opts ...SpanOption) (context.Context, Span)

	// Shutdown flushes and shuts down the tracer.
	Shutdown(ctx context.Context) error
}

// Span represents a single operation within a trace.
type Span interface {
	// End completes the span.
	End()

	// SetAttribute adds a key-value attribute to the span.
	SetAttribute(key string, value any)

	// RecordError records an error on the span.
	RecordError(err error)

	// SetStatus sets the span status.
	SetStatus(code SpanStatusCode, description string)
}

// SpanOption configures span creation.
type SpanOption func(*SpanConfig)

// SpanConfig holds span creation configuration.
type SpanConfig struct {
	// Kind specifies the span kind (client, server, producer, consumer, internal).
	Kind SpanKind
}

// SpanKind identifies the role of a span in a trace.
type SpanKind int

const (
	// SpanKindInternal is the default span kind, for work internal to a service.
	SpanKindInternal SpanKind = iota
	// SpanKindServer marks a span handling an inbound request.
	SpanKindServer
	// SpanKindClient marks a span issuing an outbound request.
	SpanKindClient
	// SpanKindProducer marks a span publishing a message to a broker.
	SpanKindProducer
	// SpanKindConsumer marks a span consuming a message from a broker.
	SpanKindConsumer
)

// SpanStatusCode represents the status of a span.
type SpanStatusCode int

const (
	// SpanStatusUnset is the default status, leaving the span outcome unspecified.
	SpanStatusUnset SpanStatusCode = iota
	// SpanStatusOK marks the span's operation as successful.
	SpanStatusOK
	// SpanStatusError marks the span's operation as failed.
	SpanStatusError
)

// WithSpanKind returns a SpanOption that sets the span kind.
func WithSpanKind(kind SpanKind) SpanOption {
	return func(cfg *SpanConfig) {
		cfg.Kind = kind
	}
}
